### PR TITLE
feat(azure-ephemeral-disk-setup): introduce experimental service for managing ephemeral disks

### PR DESCRIPTION
**Pull request by cjp256**
This commit introduces a new systemd service `azure-ephemeral-disk-setup` that automatically detects, validates, formats, and mounts Azure ephemeral disks (both NVMe and SCSI) with optional RAID-0 aggregation support. The service is designed to complement existing tools for managing SCSI resource disk like cloud-init and WALinuxAgent.

**Key Features:**
- Detects local NVMe disks by model name ("Microsoft NVMe Direct Disk" variants)
- Supports SCSI resource disk management (opt-in via configuration)
- Aggregates multiple NVMe disks into RAID-0 arrays using mdadm
- Configurable filesystem types (ext4/xfs) and mount points
- Safe validation to prevent data loss on existing formatted disks
- Persistent mounting via /etc/fstab with proper systemd integration
- Compatibility with cloud-init and WALinuxAgent

**Service Configuration:**
- Configuration file: `/etc/azure-ephemeral-disk-setup.conf`
- Service runs early in boot process with proper ordering
- Idempotent operation safe for multiple runs
- Comprehensive logging via journald

**Safety Features:**
- Only formats unpartitioned, unformatted NVMe disks
- SCSI resource disk validation (NTFS with "Temporary Storage" label)
- Conflict detection with cloud-init and WALinuxAgent
- Graceful handling for read-only fstab to support immutable images

**Package Changes:**
- New binary: `azure-ephemeral-disk-setup`
- New service: `azure-ephemeral-disk-setup.service`
- New config: `/etc/azure-ephemeral-disk-setup.conf`
- New manpage: `azure-ephemeral-disk-setup(8)`
- Added mdadm dependency for disk aggregation
- Added xfsprogs recommendation for XFS support

**Build System Updates:**
- Added shellcheck validation for shell scripts
- Updated CI/CD pipelines to test new service installation
- Comprehensive test suite with 80+ scenarios covering edge cases
- Updated packaging for Debian, RPM, and Mariner distributions

**Testing Enhancements:**
- Extended selftest validation for mount configurations
- Added service status and configuration validation
- Enhanced test infrastructure with VM reallocation testing
- Improved IMDS metadata handling and error reporting

**ephemeral-disk-setup/test-cloud-config.yaml** provides a cloud-config to demonstrate customizing config with cloud-init and seeing it take effect on provisioning boot.
 
**scripts/generate-cloud-config-ephemeral-disk-setup.sh** provides a way to test this service out on any image by installing/starting it via cloud-config.

### CODE REVIEW

File: doc/rfc-ephemeral-disk-service.md [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+ExecStart=/usr/sbin/azure-ephemeral-disk-setup.sh
+
+[Install]
+WantedBy=sysinit.target
+```
+
+## Service Configuration
+
+To allow for some flexibility for customers, we initially intend to support a number of environment variables to control behaviors:
+
```
Comment (dxyinme) [inline doc/rfc-ephemeral-disk-service.md]: here is one question about use env variables to control behaviors. pls correct me if I misunderstand.

currently, if we use marketplace image, the common way to set environment variables in azure VMSS is to pass cloud-init configuration as custom data. If the `azure-ephemeral-disk-setup.service` start before `cloud-init.service`, maybe it   is hard to pass environment variables to VM.

Comment (cjp256) [reply doc/rfc-ephemeral-disk-service.md]: I would assume the service would be disabled by default in most marketplace images... so in that case maybe something like:

```yaml
#cloud-config
write_files:
  - path: /etc/default/azure-ephemeral-disk-setup
    permissions: '0644'
    owner: root:root
    content: |
      AZURE_EPHEMERAL_DISK_SETUP_AGGREGATION="mdadm"
      AZURE_EPHEMERAL_DISK_SETUP_FS_TYPE="xfs"
      AZURE_EPHEMERAL_DISK_SETUP_MDADM_CHUNK="512K"
      AZURE_EPHEMERAL_DISK_SETUP_MDADM_NAME="azure-ephemeral-md"
      AZURE_EPHEMERAL_DISK_SETUP_MOUNT_POINT="/mnt"
      AZURE_EPHEMERAL_DISK_SETUP_SCSI_RESOURCE="false"

runcmd:
  - systemctl enable --now azure-ephemeral-disk-setup
```

However, assuming it is already enabled, I agree the service should be ordered to run after cloud-init.service - that may reduce some complexity around conflict management so that may be better for those reasons too...

Comment (cjp256) [reply doc/rfc-ephemeral-disk-service.md]: Updated the docs to move the service after cloud-init network stage

File: doc/rfc-ephemeral-disk-service.md [hunk 2]
```diff
@@ -0,0 +1,482 @@
...
+## Conflict Detection Handling if SCSI Resource Disk Detected
+
+### Conflict A: WALinuxAgent Configured to Format the Resource Disk (NVMe or No NVMe)
+
+WALinuxAgent will check if the resource disk is mounted already and skip setup. However, if the NVMe disks are formatted and mounted, this will not prevent WALinuxAgent from mounting over our mount point.
+
+Mitigation: check if `/etc/waagent.conf`contains `ResourceDisk.Format=y`.
+
+```bash
+grep -q '^ResourceDisk\.Format=y' /etc/waagent.conf
```
Comment (nmeyerhans) [inline doc/rfc-ephemeral-disk-service.md]: Do we need to handle possible whitespace around the `=` character? Is there any other possible path that could contain this setting?

File: doc/rfc-ephemeral-disk-service.md [hunk 3]
```diff
@@ -0,0 +1,482 @@
...
+
+  * This service runs after `local-fs.target`, so these disks should be actively mounted if in use by another mechanism.
+
+* `mdadm --create ...` and `mkfs.<fs_type>` will be invoked without `--force` flags to ensure they fail if the disks are partitioned/formatted or in-use
+
+If any of these checks or tools fail, the failure will be reported and service will exit with failure.
+
+In the worst case if there are cases where these tools don't detect in-use disks due to unforseen conditions, the actual risk to a user is quite low as these disks are expected to be ephemeral and this process occcurs early enough in boot to avoid negatively impacting running services in the worst case.
+
+The user must assume responsibility if they are using the disks in an unexpected manner that is impossible to detect, e.g. doing direct block read/writes without a filesystem mounted.
```
Comment (nmeyerhans) [inline doc/rfc-ephemeral-disk-service.md]: I'm a little uncomfortable with this statement. In my experience, the "expected" behavior is that I, as the VM owner, own the disks, their provisioning, and their content. Admins have not historically needed to suppress system services in order to take ownership of their disks.

File: doc/rfc-ephemeral-disk-service.md [hunk 4]
```diff
@@ -0,0 +1,482 @@
...
+### Long-term Plan: First-class Configuration in cloud-init
+
+We will independently work on a toggle for cloud-init to disable SCSI resource disk handling and key off this in a future version.
+
+Example:
+
+```yaml
+datasource:
+  Azure:
+    disable_resource_disk_handling: true
```
Comment (nmeyerhans) [inline doc/rfc-ephemeral-disk-service.md]: Stylistic suggestion: Make the key be the affirmative case, and have it default to true. Admin would set it to `false` to suppress disk setup. I'd suggest something like
```yaml
datasource:
   Azure:
      resource_disk_setup: false
```

Comment (cjp256) [reply doc/rfc-ephemeral-disk-service.md]: good idea, done!

**Comment by nmeyerhans** [comment]
Might it be better to implement this functionality in `cloud-init`, which already has some disk/partition management logic, rather than introduce a standalone service?

File: doc/rfc-ephemeral-disk-service.md (hunk 3 above)
Comment (cjp256) [reply doc/rfc-ephemeral-disk-service.md]: Fair point! How about something like:

Given the limits on what we can reasonably detect, documentation will advise users not to opt-in to this service if using disks in a manner that would conflict and go undetected, e.g. doing direct block read/writes without a filesystem mounted.

Comment (nmeyerhans) [reply doc/rfc-ephemeral-disk-service.md]: If the whole thing is **opt-in**, then my concerns are moot. It's really only if it gets turned on by default, requiring an opt-**out** that I have issues. It wasn't entirely clear to me which one it will be in this proposal.

**Comment by cjp256** [comment]
> Might it be better to implement this functionality in `cloud-init`, which already has some disk/partition management logic, rather than introduce a standalone service?

We will document how to do this with cloud-init.  But there is a lot of complexity around doing this generically.  I don't see how we can use cloud-init without a runcmd or custom script (via `write_files`) to deal with the complexity.

Cloud-init also doesn't have a great way to deal with the fresh disk problem, and the Azure data source has a bit of logic to check on every activation: https://github.com/canonical/cloud-init/blob/eb9c1239d339094a4a75a4d32ab5d970d7f71215/cloudinit/sources/DataSourceAzure.py#L1562

To do this safely, it depends on the fact that the resource disk is partitioned and formatted NTFS (and empty save one file) which requires mounting, etc.

I think it's feasible to work on extending cloud-init to support better options to enable this, but that doesn't help those not using cloud-init and will just require the same logic to be duplicated multiple places.

**Comment by nmeyerhans** [comment]
> We will document how to do this with cloud-init. But there is a lot of complexity around doing this generically. I don't see how we can use cloud-init without a runcmd or custom script (via `write_files`) to deal with the complexity.

Right, I meant within the cloud-init code itself, e.g. by extending the existing disk_setup and mounts modules, and/or introducing new modules to handle the raid parts. The advantage here is that cloud-init already provides a good framework for configuration schema, tests, etc and is already well integrated with the boot process of common distros.

> Cloud-init also doesn't have a great way to deal with the fresh disk problem, and the Azure data source has a bit of logic to check on every activation: https://github.com/canonical/cloud-init/blob/eb9c1239d339094a4a75a4d32ab5d970d7f71215/cloudinit/sources/DataSourceAzure.py#L1562
> 
> To do this safely, it depends on the fact that the resource disk is partitioned and formatted NTFS (and empty save one file) which requires mounting, etc.

Why do we not need similar logic in your proposal?

> I think it's feasible to work on extending cloud-init to support better options to enable this, but that doesn't help those not using cloud-init and will just require the same logic to be duplicated multiple places.

Do we have a sense of how many people are *not* running cloud-init? I'd expect it to be a small number of edge case users, most of whom would likely want more fine-grained control over their disk setup than what's discussed here anyway. But I could be wrong.

File: doc/rfc-ephemeral-disk-service.md (hunk 3 above)
Comment (cjp256) [reply doc/rfc-ephemeral-disk-service.md]: Whoops, I realize I did lose that blurb somewhere along the way,  I updated it to be really clear up front.

This is definitely going to be disabled by default.

**Comment by cjp256** [comment]
> 
> Why do we not need similar logic in your proposal?

There is and I hope it's pretty clear?  We'll have basic checks to validate the mounts, there is no fs on the disks and we will avoid --force style flags to avoid overwriting disks that are detectably in use or formatted.  I'm just noting it because it's always enabled and it's purposely overwriting a filesystem. For the SCSI resource disk, we'll be in the same boat for sure.

> 
> > I think it's feasible to work on extending cloud-init to support better options to enable this, but that doesn't help those not using cloud-init and will just require the same logic to be duplicated multiple places.
> 
> Do we have a sense of how many people are _not_ running cloud-init? I'd expect it to be a small number of edge case users

It's definitely a small fraction.

> most of whom would likely want more fine-grained control over their disk setup than what's discussed here anyway. But I could be wrong.

I think it's perfectly fine to have custom scripts or configs to do more complex things.  I don't think it's desirable for this service to be a replacement for those.  This is meant to provide an easy-to-enable solution to solve the general problem like wala and cloud-init do for the resource disk today.  Even if we expand options in cloud-init to make it just as easy as enabling this service, we'll have all the same knobs and be running all the same checks on boot and re-implementing it in azure-init, etc.

In my wanderings, I've encountered many custom scripts (and runcmds) to do this (regardless of cloud provider).  E.g. https://github.com/Azure/azhpc-images/blob/master/customizations/setup_nvme.sh 

FWIW I think it's a good idea to at least consider whatever we can do with cloud-init as an alternative, but the output of this exercise will be something someone can use as-is, or customize and deploy with cloud-config today.

File: doc/rfc-ephemeral-disk-service.md (hunk 2 above)
Comment (cjp256) [reply doc/rfc-ephemeral-disk-service.md]: Yes, we do, good point!

Comment (cjp256) [reply doc/rfc-ephemeral-disk-service.md]: done

File: doc/azure-ephemeral-disk-setup.md [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+After:
+
+- `cloud-init.service` (a.k.a. `cloud-init-network.service`) allows users to customize service configuration using cloud-config
+- `local-fs.target` ensures local filesystems are mounted (necessary if no cloud-init)
+
+Before:
+
+- `cloud-init.target` ensures we start before cloud-init completes which services may use to determine if system is fully configured
+- `walinuxagent.service waagent.service` ensures we start prior to walinuxagent and avoid conflicts
+- `netowrk-online.target` to ensure we start before networking is considered online
```
Comment (jeremycline) [inline doc/azure-ephemeral-disk-setup.md]:
```suggestion
- `network-online.target` to ensure we start before networking is considered online
```

File: doc/azure-ephemeral-disk-setup.md [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+```
+
+# CONFIGURATION
+
+Configuration is loaded from:
+
+* `/etc/azure-ephemeral-disk-setup.conf` (preferred)
+* Environment variables (fallback)
+
+The following environment variables are supported:
```
Comment (jeremycline) [inline doc/azure-ephemeral-disk-setup.md:38]: It looks like `AZURE_EPHEMERAL_DISK_SETUP_SYSTEMD_UNIT_TIMEOUT_SECS` is also supported. I see it's not in the sample config file as well. Is it supposed to be a "private" config, or just added late and missed here?

**Review by jeremycline — CHANGES_REQUESTED** [review]
I see one minor typo and a missing config option in the docs and example conf file.

I am not very knowledgeable about bash, and I learned a number of things while reading over this, so I can't say with great confidence that this is all right. However, I read a lot of manual pages and couldn't spot anything that didn't line up with what the documented behavior is.

File: doc/azure-ephemeral-disk-setup.md (hunk 6 above)
Comment (cjp256) [reply doc/azure-ephemeral-disk-setup.md:38]: I added it quietly to discourage configuration of it, but did it in a way that it could be overwritten if needed for a mitigation or whatever.  I don't expect it should ever need to be changed.

Comment (cjp256) [reply doc/azure-ephemeral-disk-setup.md:38]: I could document it (and leave it out of conf), but also one of those things that are subject to change and undesirable to maintain the interface for.

Comment (jeremycline) [reply doc/azure-ephemeral-disk-setup.md:38]: Sounds good to me, I agree it shouldn't need to be tweaked. Just wasn't sure if the omission was intentional.

**Review by jeremycline — APPROVED** [review]

File: doc/azure-ephemeral-disk-setup.8.in [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+If invalid values are detected, the service exits with error.
+.IP \[bu] 2
+Required tools (\f[CB]mkfs.${AZURE_EPHEMERAL_DISK_SETUP_FS_TYPE}\f[R],
+\f[CB]mdadm\f[R]) must be available.
+.SS 2. Check for Existing Mounts
+.PP
+If the mount point is already configured:
+.IP \[bu] 2
+with a matching \f[CB]/etc/fstab\f[R] line containing the comment
+\f[CB]comment=azure-ephemeral-disk-setup\f[R], the service exits with
```
Comment (anhvoms) [inline doc/azure-ephemeral-disk-setup.8.in]: consider documenting that it will re-format/raid/.... freshly allocated disk if the VM gets a new disk (deallocated/start or migration)

File: ephemeral-disk-setup/azure-ephemeral-disk-setup [hunk 8]
```diff
@@ -0,0 +1,435 @@
...
+if [[ ! "$AZURE_EPHEMERAL_DISK_SETUP_SYSTEMD_UNIT_TIMEOUT_SECS" =~ ^[0-9]+$ || "$AZURE_EPHEMERAL_DISK_SETUP_SYSTEMD_UNIT_TIMEOUT_SECS" == 0 ]]; then
+    exit_failure "AZURE_EPHEMERAL_DISK_SETUP_SYSTEMD_UNIT_TIMEOUT_SECS must be a positive integer"
+fi
+
+MKFS="mkfs.${AZURE_EPHEMERAL_DISK_SETUP_FS_TYPE}"
+if ! command -v "$MKFS" &>/dev/null; then
+    exit_failure "$MKFS is not installed and is required for formatting"
+fi
+
+if ! command -v mdadm &>/dev/null; then
```
Comment (anhvoms) [inline ephemeral-disk-setup/azure-ephemeral-disk-setup]: should we require mdadm even if the customer doesn't plan on raiding (because they're only planning on using basic v6 sku without multiple disks)

File: ephemeral-disk-setup/azure-ephemeral-disk-setup [hunk 9]
```diff
@@ -0,0 +1,435 @@
...
+        fi
+    done
+
+    log "Timed out waiting for systemd unit $unit to become active: $(systemctl status "$unit" 2>&1 || true)"
+    return 1
+}
+
+wait_for_udev_to_settle() {
+    log "Waiting for udev to settle..."
+    udevadm settle --timeout 30 || true
```
Comment (anhvoms) [inline ephemeral-disk-setup/azure-ephemeral-disk-setup]: Is 30s enough for a very big sku? I feel like we might need some retry here and allow for some control of this value or how many retries. This is probably not a problem with distros using cloud-init because by now cloud-init has already done the ephemeral networking and it has already done the udevadm settle earlier. However, it might be a problem with distros that don't use cloud-init

**Review by anhvoms — CHANGES_REQUESTED** [review]

File: ephemeral-disk-setup/azure-ephemeral-disk-setup (hunk 8 above)
Comment (cjp256) [reply ephemeral-disk-setup/azure-ephemeral-disk-setup]: after further consideration, I added AZURE_EPHEMERAL_DISK_SETUP_AGGREGATION=auto (as default) to modify behavior based on whether mdadm is installed or not

https://github.com/Azure/azure-vm-utils/pull/84/commits/2dff8e94e52dddb99aaf101b6c13a07602d98746

File: ephemeral-disk-setup/azure-ephemeral-disk-setup (hunk 9 above)
Comment (cjp256) [reply ephemeral-disk-setup/azure-ephemeral-disk-setup]: added undocumented AZURE_EPHEMERAL_DISK_SETUP_UDEVADM_SETTLE_TIMEOUT_SECS

Comment (cjp256) [reply ephemeral-disk-setup/azure-ephemeral-disk-setup]: and upped the default to 60s

File: doc/azure-ephemeral-disk-setup.8.in (hunk 7 above)
Comment (cjp256) [reply doc/azure-ephemeral-disk-setup.8.in]: done!

**Review by anhvoms — APPROVED** [review]
LGTM!
